ivhd/ivhd_grad.py

```diff
@@ -0,0 +1,104 @@
+    def __init__(
+            self,
+            n_components: int = 2,
+            nn: int = 2,
+            rn: int = 1,
+            pos_weight: float = 0.5,
+            optimizer: str = 'adam',
+            optimizer_params=None,
+            steps: int = 200,
+            epsilon: float = 1e-15,
+            re_draw_remote_neighbors: bool = False,
+            verbose: bool = False,
+    ) -> None:
+        if optimizer_params is None:
+            optimizer_params = {'lr': 0.01}
+        self.n_components = n_components
+        self.nn = nn
+        self.rn = rn
+        self.pos_weight = pos_weight
+        self.optimizer = optimizer
+        self.optimizer_params = optimizer_params
+        self.simulation_steps = steps
+        self.epsilon = epsilon
+        self.re_draw_remote_neighbors = re_draw_remote_neighbors
+        self.verbose = verbose
+
+    def transform(self, X):
+        nns = self._get_nearest_neighbors_indexes(X)
+        rns = self._get_remote_neighbors_indexes(X)
+
+        x = torch.rand(X.shape[0], self.n_components, requires_grad=True)
+        optimizer = _optimizer_mapping[self.optimizer]([x], **self.optimizer_params)
+
+        for step in range(self.simulation_steps):
+            neigborhoods = x[nns]
+            if self.re_draw_remote_neighbors:
+                rns = self._get_remote_neighbors_indexes(X)
+            remote_neigborhoods = x[rns]
+
+            xu = x.unsqueeze(1)
+            dist_emb_neighbor = ((xu - neigborhoods) ** 2 + self.epsilon).sum(dim=2).sqrt()
+            dist_emb_remote = ((xu - remote_neigborhoods) ** 2 + self.epsilon).sum(dim=2).sqrt()
+
+            del_n = 0   # desired distance between point and its nearest neighbors
+            del_r = 1   # desired distance between point and its remote neighbors
+
+            cost_emb_neighbor = ((dist_emb_neighbor - del_n) ** 2).sum()
+            cost_emb_remote = ((dist_emb_remote - del_r) ** 2).sum()
+
+            cost = 2 * (cost_emb_neighbor * self.pos_weight + cost_emb_remote * (1 - self.pos_weight))
+            if self.verbose:
+                print(f"Step: {step}\tCost: {cost}")
+
+            optimizer.zero_grad()
+            cost.backward()
+            optimizer.step()
+
+        return x.detach().numpy()
+
+    def _get_nearest_neighbors_indexes(self, X: np.ndarray) -> np.ndarray:
+        # for every point in X find indexes of its 'nn' nearest neighbors
+        knn_model = NearestNeighbors(n_neighbors=self.nn + 1)
+        knn_model.fit(X)
+        _, indices = knn_model.kneighbors(X)
+        return indices[:, 1:]
+
+    def _get_remote_neighbors_indexes(self, X: np.ndarray) -> np.ndarray:
+        # for every point in X sample indices of its 'rn' remote neighbors
+        return np.random.randint(low=0, high=X.shape[0], size=(X.shape[0], self.rn))
+
+    def __sklearn_is_fitted__(self) -> bool:
+        return True  # all calculations are performed in 'transform' method
```

Comment:
Why True is hardcoded here as a response?

Reply:
Same as in previous, non-gradient, approach - there is nothing to fit. IVHD does not have any internal parameters. You cannot fit it, and then use gained knowledge on other set. Therefore, it is always already "fitted".

Reply:
Thanks